### DISCUSSION TITLE
CYK parser: how does lark work with CYK?

### DISCUSSION THREAD

**Discussion by Oleksii-Hovderchak** [lark-parser/lark#850] [Q&A] [answered]
Has anybody tested the CYK parser?

Afaik, the grammar should be in a CNF, like this:

S -> AB | BC
A -> BA | a
B -> CC | b
C -> AB | a 

So that the string 'baaba' should be now parsed well, but I have errors right now.

Tried to find some examples on the github or some tests in your code or at least a grammar example and I failed.

Could anyone help with that? Maybe I'm missing smth, thank is advance!

**Comment by MegaIng** [comment, marked as answer]
Well, you probably don't want to use cyk in lark. It is more there for backwards compatibility.

But anyway, how are you calling Lark? what errors do you get? You at least need to change the syntax to the lark syntax.

**Reply by MegaIng** [reply]
Also, you don't need to manually write CNF I think. I think lark automatically rewrites them.

**Reply by Oleksii-Hovderchak** [reply]
So I want just to understand how I could parse CYK with lark I do the next

from lark import Lark

rules = """
S -> NP VP
VP -> VP PP
VP -> V NP
VP -> eats
PP -> P NP
NP -> Det N
NP -> she
V -> eats
P -> with
N -> fish
N -> fork
Det -> a

"""

parser = Lark(rules, parser='cyk', ambiguity='explicit')

sentence = 'she eats a fish with a fork'

if __name__ == '__main__':
    print(parser.parse(sentence).pretty())

The exception I have:
lark.exceptions.GrammarError: Expecting rule or terminal definition (missing colon), at line 2 column 3

S -> NP VP
  ^

**Reply by MegaIng** [reply]
You need to use the Lark syntax. Look at the [json tutorial](https://github.com/lark-parser/lark/blob/master/docs/json_tutorial.md) for example.

**Reply by Oleksii-Hovderchak** [reply]
Well, good point regarding rewritings.. but still want to see how it works

**Reply by Oleksii-Hovderchak** [reply]
Okay, actually I have an example for 'earley' parser
That works perfectly:

from lark import Lark

grammar = """
    sentence: noun verb noun        -> simple
            | noun verb "like" noun -> comparative

    noun: adj? NOUN
    verb: VERB
    adj: ADJ

    NOUN: "flies" | "bananas" | "fruit"
    VERB: "like" | "flies"
    ADJ: "fruit"

    %import common.WS
    %ignore WS
"""

parser = Lark(grammar, start='sentence', parser='earley', ambiguity='explicit')

sentence = 'fruit flies like bananas'

if __name__ == '__main__':
    print(parser.parse(sentence).pretty())

**Reply by Oleksii-Hovderchak** [reply]
and if I use this syntax for cyk, I would have an exception:

\lib\site-packages\lark\parsers\cyk.py", line 110, in parse
    raise ParseError('Parsing failed.')
lark.exceptions.ParseError: Parsing failed.

**Reply by MegaIng** [reply]
This is (I think) because the automatic expansion of the shorthand `adj?` syntax is `_temp: adj |`, which means that `_temp` might be empty, which is something that the `cyk` parser can't handle. You need to manually write `noun: adj NOUN | NOUN`. I would suggest not using cyk at all.

**Reply by Oleksii-Hovderchak** [reply]
Yeah, I also already understand I would not use CYK but need also to explain to my team (for example) why I would not do it. For this purpose, I want to run it successfully at least one time. But still failed
even with your suggestion regarding adj NOUN | NOUN

**Reply by Oleksii-Hovderchak** [reply]
Tried also with simple lalr grammar, same problem: ParseError
Could it be the issue?

**Reply by MegaIng** [reply]
Ok, `adj?` wasn't the problem. The problem is a terminal collision between `VERB` and the literal `"like"`. There are a few ways to resolve this, one is the following:

```
sentence: noun verb noun -> simple
        | noun verb "like" noun -> comparative

noun: adj? _noun
!adj: "fruit"
!verb: "like" | "flies"

!_noun: "flies" | "bananas" | "fruit"

%import common.WS
%ignore WS
```

**Reply by MegaIng** [reply]
But as to why not cyk:

- cyk is the least supported parser option in Lark
- cyk is the slowest option
- cyk will be removed in 1.0 (unless you have good arguments as to why not)

**Reply by Oleksii-Hovderchak** [reply]
oh thanks a lot! But why the collision started? CYK does not support like 'terminal reference'?

**Reply by Oleksii-Hovderchak** [reply]
omg I guess I got it, CYK uses the standard lexer, right? So here is where earley also wins

**Reply by MegaIng** [reply]
Yes.

**Reply by Oleksii-Hovderchak** [reply]
Great, thanks for your help!
Good luck!
